### PR TITLE
boot: observe successful command line update, provide a default 

### PR DESCRIPTION
**Pull request by bboozzoo**
Add bootstate handling for observing a successful command line update. At the
same time, provide backwards compatibility for the systems built before tracking
of current kernel command lines was introduced in modeenv and set one on
successful boot.

This is transparent for resealing until the patches plugging current kernel command lines into resealToModeenv() a posted and landed.

### CODE REVIEW

File: boot/assets.go [hunk 1]
```diff
@@ -847,3 +847,53 @@ func observeSuccessfulBootAssets(m *Modeenv) (newM *Modeenv, drop []*trackedAsse
 	}
 	return newM, drop, nil
 }
+
+func observeSuccessfulCommandLine(m *Modeenv) (*Modeenv, error) {
```
Comment (pedronis) [inline boot/assets.go:851]: these probably merit some descriptiive comments

Comment (pedronis) [inline boot/assets.go:851]: should these functions live in cmdline.go instead perhaps?

File: boot/bootstate20.go [hunk 2]
```diff
@@ -720,3 +720,44 @@ func trustedAssetsBootState(dev Device) *bootState20BootAssets {
...
+		// not using trusted boot assets, bootloader config is not
+		// managed and command line can be manipulated externally
+		return update, nil
+	}
+
+	var newM *Modeenv
+	if len(u20.modeenv.CurrentKernelCommandLines) == 0 {
+		// no command lines tracked in modeenv yet, this can happen when
+		// having booted with a newer snapd
+		newM, err = observeSuccessfulCommandLineCompatBoot(ba20.dev.Model(), u20.writeModeenv)
```
Comment (pedronis) [inline boot/bootstate20.go:748]: I see we do the same with observeSuccessfulBootAssets,  I didn't wonder before but should these take as input writeModeenv or modeenv, it's a bit unclear to me right now?  if they really take writeModeenv do they need to make a copy?

File: boot/bootstate20.go [hunk 3]
```diff
@@ -720,3 +720,44 @@ func trustedAssetsBootState(dev Device) *bootState20BootAssets {
...
+		// XXX: does this change when we expose the ability to add
+		// things to the command line?
+
+		// not using trusted boot assets, bootloader config is not
+		// managed and command line can be manipulated externally
+		return update, nil
+	}
+
+	var newM *Modeenv
+	if len(u20.modeenv.CurrentKernelCommandLines) == 0 {
```
Comment (pedronis) [inline boot/bootstate20.go:745]: maybe this if should be inside a single observe* helper that takes this decision inside?

**Review by pedronis — COMMENTED** [review]
overall looks good, but some comments/questions

File: tests/nested/core20/tpm/task.yaml [hunk 4]
```diff
@@ -52,3 +52,17 @@ execute: |
...
+    cmdline="$(nested_exec "cat /proc/cmdline")"
+    test "$cmdline" = "$cmdline_modeenv"
+    echo "$cmdline" | MATCH "snapd_recovery_mode=run "
+
+    # check that kernel command lines is restored when booted successfully
+    nested_exec "sudo sed -i -e 's/current_kernel_command_lines=.*/current_kernel_command_lines=/' /var/lib/snapd/modeenv"
+    boot_id="$( nested_get_boot_id )"
+    nested_exec "sudo reboot" || true
+    nested_wait_for_reboot "${boot_id}"
+    cmdline_modeenv_restored="$(grep current_kernel_command_lines= < modeenv  | sed -e 's/current_kernel_command_lines=//' | jq -r '.[0]')"
```
Comment (anonymouse64) [inline tests/nested/core20/tpm/task.yaml:67]: this check is wrong, we don't modify the `$PWD/modenv` file here from above, so when we the sed we do it on `/var/lib/snapd/modeenv`, and don't modify `$PWD/modeenv` that is being used here. I think this needs to be doing something like

```suggestion
    cmdline_modeenv_restored="$(grep current_kernel_command_lines= < /var/lib/snapd/modeenv  | sed -e 's/current_kernel_command_lines=//' | jq -r '.[0]')"
```

**Review by anonymouse64 — CHANGES_REQUESTED** [review]
comment on the spread test

**Review by anonymouse64 — COMMENTED** [review]
lgtm, but see my previous comment about the spread test

File: boot/bootstate20.go [hunk 5]
```diff
@@ -720,3 +720,44 @@ func trustedAssetsBootState(dev Device) *bootState20BootAssets {
...
 	}
 }
+
+// bootState20CommandLine implements the successfulBootState interface for
+// kernel command line
+type bootState20CommandLine struct {
+	dev Device
+}
+
+func (ba20 *bootState20CommandLine) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
```
[the remark this answers is not in the document]
Comment (pedronis) [reply boot/bootstate20.go:730]: I prefer the split, it matches different entries in the modeenv for each, that's also why I asked about what the input should be for the observe* helpers given that the operate on orthogonal input (but not strong opinion on that)

File: boot/assets.go (hunk 1 above)
Comment (bboozzoo) [reply boot/assets.go:851]: added

Comment (bboozzoo) [reply boot/assets.go:851]: and moved

File: boot/bootstate20.go (hunk 3 above)
Comment (bboozzoo) [reply boot/bootstate20.go]: moved to a single func

File: tests/nested/core20/tpm/task.yaml (hunk 4 above)
Comment (bboozzoo) [reply tests/nested/core20/tpm/task.yaml]: updated

File: boot/cmdline.go [hunk 6]
```diff
@@ -169,3 +169,76 @@ func ComposeCandidateCommandLine(model *asserts.Model) (string, error) {
...
+		return m, nil
+	default:
+		return observeSuccessfulCommandLineUpdate(m)
+	}
+}
+
+// observeSuccessfulCommandLineUpdate observes a successful boot with a command
+// line which is expected to be listed among the current kernel command line
+// entries carried in the modeenv. One of those entries must match the current
+// kernel command line of a running system.
```
Comment (pedronis) [inline boot/cmdline.go:199]: ... and will be recorded alone as in use.

**Review by pedronis — APPROVED** [review]
thank you

File: boot/cmdline.go [hunk 7]
```diff
@@ -169,3 +169,76 @@ func ComposeCandidateCommandLine(model *asserts.Model) (string, error) {
...
+		return nil, err
+	}
+
+	// get the current command line
+	cmdlineBootedWith, err := osutil.KernelCommandLine()
+	if err != nil {
+		return nil, err
+	}
+	if !strutil.ListContains([]string(m.CurrentKernelCommandLines), cmdlineBootedWith) {
+		return nil, fmt.Errorf("internal error: current command line content %q not matching any expected entry",
```
Comment (anonymouse64) [inline boot/cmdline.go:212]: mmm thinking about it, I don't think this is an internal error, so maybe just drop the `internal error:` prefix here maybe

File: tests/nested/core20/tpm/task.yaml [hunk 8]
```diff
@@ -58,3 +58,15 @@ execute: |
...
+
+    # check that kernel command lines is restored when booted successfully
+    nested_exec "sudo sed -i -e 's/current_kernel_command_lines=.*/current_kernel_command_lines=/' /var/lib/snapd/modeenv"
+    boot_id="$( nested_get_boot_id )"
+    nested_exec "sudo reboot" || true
+    nested_wait_for_reboot "${boot_id}"
+    # the test will race with snapd updating the modeenv, wait for the modenv to
+    # have a good state
+    nested_retry_until_success 10 5 "grep -E \"current_kernel_command_lines=.*snapd_recovery_mode=run\" < /var/lib/snapd/modeenv"
+    nested_exec "cat /var/lib/snapd/modeenv" > modeenv.after-reboot
```
Comment (anonymouse64) [inline tests/nested/core20/tpm/task.yaml:70]: thanks for this

File: boot/bootstate20.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (anonymouse64) [reply boot/bootstate20.go:730]: sure, that seems fine to me, just wondering

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for the changes

**Pull request merged**